### ISSUE TITLE
Should Pkg.clone/update initialize git submodules?

### ISSUE THREAD

**Issue by wavexx** [JuliaLang/julia#11586]
In Polyglot.jl I'm sharing driver data (which is just source) as a git submodule in ``src/drivers``. In ``deps/build.jl`` I'm running ``git submodule update --init`` to fetch the source, but this is not portable (no clue if this would work under windows) and somehow ugly. I noticed the package ``HttpParser`` does the same, I don't know about possible others.

I know subtrees would be an option, but I'd like to avoid that. Is it reasonable that submodules should be initialized/updated if being used by the package?

**Comment by tkelman** [comment]
No, I don't think we should encourage submodules. You can ```run(`git submodule update --init`)``` right now and it should work fine, shelling out to git is how the package manager does everything right now anyway. That may change in the not-too-distant future ref #11196, but there we can probably expose libgit2's submodule handling functionality with a separate API.

**Comment by wildart** [comment]
I do not think it would be a priority issue if ever.

**Comment by tkelman** [comment]
While I don't think we should go out of our way to incorporate submodule-handling features into Pkg proper, if exposing and documenting the libgit2 API's for working with submodules isn't too much trouble I think we should try to do so - otherwise removing command-line git from the binaries would result in a regression in available functionality here.

**Comment by wildart** [comment]
I think that git tools should stay in binaries (we talk about Windows), because libgit2 functionality couldn't be compared to command-line git. Developers would require git even if the package system will work without it. libgit2 is not substitute for git.

**Comment by tkelman** [comment]
No, I strongly disagree. Getting rid of git will cut down the size of the binaries dramatically. I don't want to bundle them if Pkg isn't going to use them any more. libgit2 is supposed to be a substitute for git as far as Pkg is concerned, that's the whole point.

**Comment by wildart** [comment]
Well, `git` could be removed from binary, but developers will need it anyway. It is not that `libgit2` could not do routine operations (merging is still problematic), but it is not a substitute for command line git tools.   Some common `git` operations could be really complicated from libgit2 point of view.
Package manager works with packages and their versions. Any changes, smaller than a version, to a package considered to be handled by the user. Thus, package manager does not have commit or reset functionality, and it shouldn't. In any case, Windows users would need binary `git` tools installed to be able to develop Julia packages.

**Comment by tkelman** [comment]
I think you're missing my point.

> we can probably expose libgit2's submodule handling functionality with a separate API

Is there any good reason to not do this?

> Thus, package manager does not have commit or reset functionality, and it shouldn't.

Why not? I don't see why we couldn't conceptually replace running git commands from shell mode while inside a package directory with calls to libgit2 functions from julia mode in the repl. While libgit2 may not be "a substitute for command line git tools," we should really endeavor to make it as close as possible via the Julia bindings. Julia's C FFI should allow for getting pretty close to a complete replacement for `git` with nearly the same usability. Why couldn't there be a `Pkg.commit` or `Pkg.reset` for example?

> Windows users would need binary git tools installed to be able to develop Julia packages.

Depending on how much of the above is possible, this is probably the case. And for at least a decent number of windows users, they'll likely be more comfortable with one of the various gui's to git rather than the command line. That doesn't change anything about this issue though.

**Comment by wildart** [comment]
The reason is "do not reinvent bicycle". It would take a lot of effort to reimplement command-line git tools in Julia/libgit2, and I do not see a point in this exercise. We using libgit2 in the package manager to streamline performance, and only performance, not user experience.

**Comment by tkelman** [comment]
> I do not see a point in this exercise

To remove a major GPL dependency, which is large and problematic for a number of reasons.

> It would take a lot of effort to reimplement command-line git tools 

Like what? Don't you have to implement commit anyway in order to do `Pkg.tag` and `Pkg.publish` properly? How hard would reset or submodule handling be, really?

> We using libgit2 in the package manager to streamline performance, and only performance, not user experience.

I don't think this is the right viewpoint here. Let's improve both, and remove an inelegant dependency while we're at it.

**Comment by wildart** [comment]
Don't get it wrong, I'm completely in favor of removing the command-line git tool dependency from the Julia codebase. No question here. A large portion of git functionality will be available under a package manager "hood" through libgit2. So, anybody could use it to manage git repos. However, I strongly disagree to rewrite complete set of git tools in Julia/libigt2. There are plenty of tools to work with git repositories, why bother to create another subpar implementation which it would be considering libgit2 functional incompleteness. I see it as a matter of codebase support, libgit2 is still under development. Every version introduce plenty of changes. Look at LibGit2.jl it was written to support libgit2 0.21, and the package is already incompatible with 0.22. So, I would prefer to focus on functionality that is required for the package manager rather on complete set of git tools.
P.S. While writing all this, I've got an idea about git support in Julia. A correct way to introduce git support in Julia would be through a git REPL mode rather than through package manager commands. Dedicated lightweight repository management REPL mode could be a solution which can be developed in parallel and contribute to the libgit2 support in the Julia codebase.

**Comment by tkelman** [comment]
That makes sense. There are sort of 3 levels here - core functionality for Pkg, nice-to-haves, and features that would be too much work to expose in Julia via libgit2 bindings. For the sake of this issue, where would submodule init and update lie on that spectrum? You're right that the most complicated features like full interactive rebase, bisect, that kind of thing are probably not worth the effort to do via libgit2 for now.

A git REPL mode sounds like a pretty cool idea.

**Comment by wildart** [comment]
I would put "submodules" somewhere between nice-to-have and too-much-work. I haven't look in detail what is libgit2 submodule functionality, but from a first glance - 34 functions dedicated for submodule handling which is extensive. Does it worth introduce submodules? Maybe. libgit2 coverage looks complete on this issue. Does it need to be added to Pkg? Maybe. How common is usage of the submodules? Maybe other options can be explored, considering that submoduled repo can be handled in `deps/build.jl` script.
I looked at @wavexx package, it has tight coupling with another repo, so "submodules" is natural way. I guess we could put "submodules" to wishlist.

**Comment by tkelman** [comment]
I'm now convinced we don't need this in base. There's still a LibGit2.jl package, and instead of making it useless after libgit2 gets included in base, we can instead use it for any advanced functionality like submodules that aren't absolutely required by Pkg.
